src/preact/use-in-view.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useEffect, useLayoutEffect, useRef, useState} from './index';
+
+/**
+ * @param {HTMLElement} node
+ * @param {function():(function():undefined|undefined)} effect
+ * @param {!Array<*>=} opt_deps
+ * @return {object} ref
+ */
+export function useInViewEffect(node, effect, opt_deps) {
```

Comment:
I think there are a few issues with this. Let me redraw this method and let's discuss pros/cons. I will also comment on the body of this hook below to stress where we will have issues. Note, that this fragment also has issues, but fewer and, once we get a handle of this, we can see what we can address those.

The first important point: the first argument should be a `Ref` and not a node. Otherwise we will miss the first update. I will further qualify it at the call site. For now let's assume there's the mentioned `useRefEffect`. Let's see how it works and we will get to its implementation later.

```
export function useInViewEffect(ref, effect, opt_deps) {
  const isIntersectingRef = useRef(false);
  const unsubscribeRef = useRef(null);
  useRefEffect(
    ref,
    (current) => {
        const observer = new IntersectionObserver(entries => {
          const {isIntersecting} = entries[entries.length - 1];
          if (isIntersecting !== isIntersectingRef.current) {
            isIntersectingRef.current = isIntersecting;
            if (unsubscribeRef.current) {
              unsubscribeRef.current();
              unsubscribeRef.current = null;
            }
            if (isInersecting) {
              unsubscribeRef.current = effect();
            }
          }
        });
        observer.observe(current);
        return () => {
            if (unsubscribeRef.current) {
              unsubscribeRef.current();
            }
            isIntersectingRef.current = false;
            unsubscribeRef.current = null;
            observer.disconnect();
        };
    },
    opt_deps
  );
}
```

It's slightly longer, but more universal and it does quite a bit more. It can also be simplified a bit.

Reply:
Dima, thanks for providing this valuable feedback!

Two notes:
- In this code snippet, the `if (unsubscribeRef.current) {...}` blocks are unreachable because the `effect` passed in by timeago doesn't have a `return` value. Because of this I removed 
- The `IntersectionObserver` is instantiated every time the effect is called. IIUC the tradeoff we are making here is that there is additional cost instiatiating it every time, but there will be no leakage in terms of disconnecting unintended nodes that may have been observed through other means (?)

Reply:
/1/ It's ok for a specific effect to not have a cleanup task. But the universal effect should expect it nonetheless. That's why we do `if (unsubscribeRef.current) {...}`

/2/ The `opt_deps` controls how often the `IntersectionObserver` is created. We can always default `opt_deps` to an empty array. E.g. `opt_deps || []`.

/3/ I think we really have two effects here: `useIntersectionEffect()` and `useInViewEffect()`. The second calls the first. Not a huge difference, but there might be some good optimizations possible.

Reply:
I see your point -- the `unsubscribeRef` is not used in this case but present for completeness.

@jridgewell and I discussed some this afternoon and reached the resolution that the original approach with `useIsIntersecting` is likely more preferable in this case. It appears to be both more idiomatic and much cleaner.

PTAL at the recent changes which should reflect the benefits of both approaches:
- Provide a generic `useIsIntersecting` which sets and returns `isIntersecting` as the only piece of state being changed, avoiding unnecessary rerenders in a single sequence. This custom hook can be used in the `timeago` use case as well as more generically, i.e. for `amp-animation` or video management.
- `useIsIntersecting` is bundled such that it instantiates the `IntersectionObserver` once per component and shares it at that level. To that effect, the observer will `unobserve` rather than `disconnect` in the cleanup phase leading to better symmetry and lower risk of state leakage.
- Finally, although for the timeago use case the `useIsIntersecting` will trigger the render phase on enter _and_ exit viewport, the `timestampRef` will be the source of truth for the timestamp value that only updates on _enter_ viewport. While on exit the render phase will still perform the DOM diff on the element, the component will not actually repaint on the page.

In this case we alleviate the concern around a [3x rerender in a single sequence](https://github.com/ampproject/amphtml/pull/26507#discussion_r379161030) with a solution that will be more universally usable in our other component use cases.

Reply:
I caught up with @jridgewell as well. But I don't entirely agree. I think we can proceed this way, however, provided we fix all the issues with the `useIsIntersecting` implementation.

But to explain, I see it this way. IMHO, the most idiomatic approach here is:

```
function TimeAgo(props) {
  const [timestamp, setTimestamp] = useState('');
  const ref = useRef(null);

  useEffect(() => {
    const observer = new IntersectionObserver((entries) => {
      const last = entries[entries.length - 1];
      if (last.isIntersecting) {
        setState(true);
      }
    });
    observer.observe(ref.current);
    return () => observer.disconnect();
  }, []);  

  return <time ref={ref} datetime={datetime}>{timestamp}</time>;
}
```

There are no extra rerendering here; all state is accounted for and represents the logical component state; the boundary between rendering and side-effect is clear as well. For some reason we may want to hide the use of `IntersectionObserver` here, which could be valid (although maybe premature?). But I don't think simply hiding InOb should drastically change the structure of this component.

To drive a point a bit deeper, there's a bug in the following fragment:
```
  if (isIntersecting === null || isIntersecting) {
    // Only update DOM on prop mutation or enter viewport
    timestampRef.current = getFuzzyTimestampValue(props);
  }
```

This will change the timestamp value on each rerender once `isIntersecting==true`. As opposed to "only change once each time the element enters viewport". With my code above, this kind of bug is much harder to make.

Reply:
Expanded on your approach with the following changes:
- Replaced `setState(true)` with `setTimestamp(...)` since there is neither  an `[isIntersecting, setState]` pair initialized nor usage of something like `isIntersecting`.
- Passed in `props` dependency due to above -- this also re-introduces the "bug" you described which I will discuss later
- Non-JSX formatting since it's currently not supported.

```
export function Timeago(props) {
  const {0: timestamp, 1: setTimestamp} = useState('');
  const ref = useRef(null);

  useEffect(() => {
    const observer = new IntersectionObserver(entries => {
      const last = entries[entries.length - 1];
      if (last.isIntersecting) {
        setTimestamp(getFuzzyTimestampValue(props));
      }
    });
    observer.observe(ref.current);
    return () => observer.disconnect();
  }, [props]);

  return createElement('time', {datetime: props['datetime'], ref}, timestamp);
}
```

However, both your approach and my modifications to it does away with two abstractions that I thought were a priority in the discussion so far: 
(1) A shared, reusable IntersectionObserver across timeago instances on a single document
(2) A generic `useIsIntersecting` hook that many components can use to perform logic based on enter/exit viewport events. It seems with the current approach, most components will define near-identical effects in their own logic.

It is also worth noting that we have a double render on component instantiation due to `ref` being passed in as a prop -- it appears to render once on instantiation with `useRef(null)` and a second time when its `current` is attributed to the DOM node. Normally mutating refs itself shouldn't trigger a render, but in this case it does because it mutates the element props being passed into the `return`. 

Finally, I would argue that the code snippet above wherein render occurs "anytime it is intersecting" rather than "only when entering viewport" is not a bug because it intentionally handles prop mutations, i.e. via amp-bind.

Reply:
Let's start with the following statement, b/c I think it's very important to clarify it first:

> (1) A shared, reusable IntersectionObserver across timeago instances on a single document

We cannot make a reusable `IntersectionObserver` using a `ref` in the hook. LMK if you want to go over this offline. But the main caveat there is that the `useRef()`, `useState()`, `useMemo()`, etc - all these hooks reserve the storage space in the scope of a specific component they are "declared" within. If we want to make an `IntersectionObserver` instance sharable, we need to use a context provider.